### PR TITLE
crimson/os/seastore: generalize journal tail calculations with CircularBoundedJournal

### PR DESCRIPTION
**Pull request by cyx1231st**
This PR contains the unmerged part of https://github.com/ceph/ceph/pull/46725 (last 2 commits), which also enables test_transaction_manager with CBJournal trimming.

~~It also shares cleanup commits with https://github.com/ceph/ceph/pull/47646 (first 3 commits)~~

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/async_cleaner.h [hunk 1]
```diff
@@ -1207,6 +1194,9 @@ class AsyncCleaner : public SegmentProvider, public JournalTrimmer {
    * Segments calculations
    */
   std::size_t get_segments_in_journal() const {
+    if (journal_type == journal_type_t::CIRCULAR) {
+      return 0;
+    }
```
Comment (cyx1231st) [inline src/crimson/os/seastore/async_cleaner.h:1199]: @myoungwon The root issue is that there are no segments in journal with CircularBoundedJournal.

Comment (myoungwon) [reply src/crimson/os/seastore/async_cleaner.h:1199]: Yeah, I originally added this line before, but your comment is that I need to put this kind of condition to more general place. Anyway, this is ok for now.

Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.h:1199]: Sorry, I'm not aware of the exact issue until I tested it myself.

Comment (myoungwon) [reply src/crimson/os/seastore/async_cleaner.h:1199]: No worries. Anyway, I need to time to look at this if my review is needed.

Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.h:1199]: Reviews are always valuable!

Feel free to also comment wherever the intention is not clear, so I can elaborate.

File: src/crimson/os/seastore/seastore_types.cc [hunk 2]
```diff
@@ -129,6 +129,81 @@ std::ostream &operator<<(std::ostream &out, const paddr_t &rhs)
...
+  assert(roll_size > 0);
+
+  segment_seq_t jseq = segment_seq;
+  seastore_off_t joff;
+  if (type == journal_type_t::SEGMENTED) {
+    joff = offset.as_seg_paddr().get_segment_off();
+  } else {
+    assert(type == journal_type_t::CIRCULAR);
+    auto boff = offset.as_blk_paddr().get_block_off();
+    assert(boff <= MAX_SEG_OFF);
```
Comment (myoungwon) [inline src/crimson/os/seastore/seastore_types.cc:150]: BLOCK_OFF_MAX ?

File: src/crimson/os/seastore/seastore_types.cc [hunk 3]
```diff
@@ -129,6 +129,81 @@ std::ostream &operator<<(std::ostream &out, const paddr_t &rhs)
...
+  if (off >= 0) {
+    jseq += (off / roll_size);
+    joff += (off % roll_size);
+    if (joff >= roll_end) {
+      ++jseq;
+      joff -= roll_size;
+    }
+  } else {
+    auto mod = static_cast<segment_seq_t>((-off) / roll_size);
+    joff -= ((-off) % roll_size);
```
Comment (myoungwon) [inline src/crimson/os/seastore/seastore_types.cc:166]: If offset is blk_paddr_t, we probably need to use int64_t, not int32_t.

File: src/crimson/os/seastore/seastore_types.cc [hunk 4]
```diff
@@ -129,6 +129,81 @@ std::ostream &operator<<(std::ostream &out, const paddr_t &rhs)
...
+  assert(roll_size > 0);
+
+  int64_t ret = static_cast<int64_t>(segment_seq) - r.segment_seq;
+  ret *= roll_size;
+  if (type == journal_type_t::SEGMENTED) {
+    ret += (static_cast<int64_t>(offset.as_seg_paddr().get_segment_off()) -
+            static_cast<int64_t>(r.offset.as_seg_paddr().get_segment_off()));
+  } else {
+    assert(type == journal_type_t::CIRCULAR);
+    ret += (static_cast<int64_t>(offset.as_blk_paddr().get_block_off()) -
```
Comment (myoungwon) [inline src/crimson/os/seastore/seastore_types.cc:200]: Would we need to consider the following case (journal_head is behind alloc_head)?

                    Journal_head                          alloc_head
journal begin        V                                            V           journal end

**Comment by cyx1231st** [comment]
```
unittest-transaction-manager: ../src/crimson/os/seastore/cache.cc:1773: crimson::os::seastore::Cache::replay_delta(crimson::os::seastore::journal_seq_t, crimson::os::seastore::paddr_t, const crimson::os::seastore::delta_info_t&, const crimson::os::seastore::journal_seq_t&, const crimson::os::seastore::journal_seq_t&, crimson::os::seastore::sea_time_point)::<lambda(auto:178)> [with auto:178 = b
oost::intrusive_ptr<crimson::os::seastore::CachedExtent>]: Assertion `extent->last_committed_crc == delta.prev_crc' failed.
```

File: src/crimson/os/seastore/seastore_types.cc (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.cc:200]: There is no alloc_head/dirty_head in the design, probably you're refering to alloc_tail/dirty_tail?

Like dirty_tail, journal_head is always ahead of alloc_tail.

File: src/crimson/os/seastore/seastore_types.cc (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.cc:166]: If 2GB is not enough for journaling with RBM, I think we will need a separate PR to address this by introducing the original int32_t segment_off_t back, use it where appropriate, and then extend seastore_off_t to int64_t.


Refer to the commit https://github.com/ceph/ceph/commit/c93d0b7b1e7be2e3879ec384a951edb247a9f775

File: src/crimson/os/seastore/seastore_types.cc (hunk 2 above)
Comment (myoungwon) [reply src/crimson/os/seastore/seastore_types.cc:150]: get_block_off() can return the maximum of 56 bit value. On the other hand, MAX_SEG_OFF is int32_t. Therefore, this assert can be triggered even though the blk_paddr_t has correct offset. Am I wrong?

Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.cc:150]: The specific assert is necessary to detect the potential overflow with the current `seastore_off_t`.

See https://github.com/ceph/ceph/pull/47698#discussion_r951039417, I think this is a separate issue. Before that, I'll assert that the related configurations are in the reasonable ranges.

For example https://github.com/ceph/ceph/blob/d9526dc2a380fd91c3316eac1906cd97ad244a64/src/crimson/os/seastore/async_cleaner.h#L568-L570

Is it reasonable to you?

File: src/crimson/os/seastore/seastore_types.cc (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.cc:200]: > journal_head is always ahead of or equal to alloc_tail

I meant alloc_tail/dirty_tail are of type `journal_seq_t` that includes `segment_seq` for the comparison.

> Would we need to consider the following case (journal_head is behind alloc_head)?

I think you mean journal_head.offset is behind alloc_head.offset. This is already considerred with
```
int64_t ret = static_cast<int64_t>(segment_seq) - r.segment_seq;
ret *= roll_size;
```

File: src/crimson/os/seastore/seastore_types.cc (hunk 2 above)
Comment (myoungwon) [reply src/crimson/os/seastore/seastore_types.cc:150]: OK

File: src/crimson/os/seastore/seastore_types.cc (hunk 4 above)
Comment (myoungwon) [reply src/crimson/os/seastore/seastore_types.cc:200]: Understood.

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
Besides the int32_t seastore_off_t and the missing paddr->journal_seq_t map, is the change good to you? @myoungwon @athanatos

**Comment by cyx1231st** [comment]
jenkins test make check

**Review by myoungwon — APPROVED** [review]
paddr->journal_seq_t mapping to address out-dated delta is the next PR?

**Comment by cyx1231st** [comment]
> paddr->journal_seq_t mapping to address out-dated delta is the next PR?

No, I'm working towards splitting trimming out of the segment-reclaim logic. Feel free to take it if want to.

**Review by athanatos — APPROVED** [review]

**Comment by myoungwon** [comment]
> No, I'm working towards splitting trimming out of the segment-reclaim logic. Feel free to take it if want to.

Ok. I'll add it to my to-do-list. So, I'm currently working on 1) prototyping RBM OOL path, 2) RBM cleaner. 3) paddr->journal_seq_t mapping.

**Comment by cyx1231st** [comment]
Make sense to me.

> 1) prototyping RBM OOL path

I'm thinking myself that the current segmented ool writes might be able to be organized by raw blocks without record metadata.

So, in another word, it will be always benefitial to think about how the RBM OOL path can work based on the existing mechanisms, with less wheels needed to invent.

> 2) RBM cleaner

I'll work on introducing the initial general interface from the cleaner to EPM, so we can have a baseline to understand how to integrate RBCleaner to EPM properly.

**Pull request merged**